File: src/my_class_pkg/scripts/action_server.py
#!/usr/bin/env python3
import rospy
import actionlib
from my_class_pkg.msg import MyActionMsgAction, MyActionMsgFeedback, MyActionMsgResult

class CountActionServer:
    def __init__(self):
        # 创建动作服务器（服务名：my_action）
        self.server = actionlib.SimpleActionServer(
            "my_action", MyActionMsgAction, self.execute, False
        )
        self.server.start()
        rospy.loginfo("Python Action Server started, waiting for goal...")

    def execute(self, goal):
        rate = rospy.Rate(1) # 1Hz（每秒计数1次）
        target = goal.target
        rospy.loginfo(f"Action server started: counting up to {target}")

        # 修复核心错误：判断逻辑写反了！
        # 原错误：if self.server.is_preempt_requested() or not rospy.is_shutdown():
        # 正确逻辑：只有当「客户端取消请求」OR「ROS关闭」时，才终止
        for i in range(target + 1):
            if self.server.is_preempt_requested() or rospy.is_shutdown():
                self.server.set_preempted()
                return

            # 发布进度反馈（核心：逐秒返回计数进度）
            feedback = MyActionMsgFeedback()
            feedback.current = i
            feedback.progress = float(i) / target if target != 0 else 1.0
            self.server.publish_feedback(feedback)
            rospy.loginfo(f"Progress: current = {i}, progress = {feedback.progress:.2f}")
            rate.sleep()

        # 执行完成，返回成功结果
        result = MyActionMsgResult()
        result.success = True
        self.server.set_succeeded(result)
        rospy.loginfo("Action finished successfully!")

if __name__ == "__main__":
    rospy.init_node("python_action_server")
    server = CountActionServer()
    rospy.spin()

File: src/my_class_pkg/src/action_server.cpp
#include "ros/ros.h"
#include "actionlib/server/simple_action_server.h"
#include "my_class_pkg/MyActionMsgAction.h"

// 定义动作服务器类型
typedef actionlib::SimpleActionServer<my_class_pkg::MyActionMsgAction> Server;

// 动作回调函数：处理目标请求
void execute(const my_class_pkg::MyActionMsgGoalConstPtr& goal, Server* as) {
    ros::Rate rate(1); // 1Hz（每秒执行一次）
    my_class_pkg::MyActionMsgFeedback feedback; // 进度反馈
    my_class_pkg::MyActionMsgResult result;     // 最终结果

    int target = goal->target;
    ROS_INFO("Action server started: counting up to %d", target);

    // 模拟计数过程，实时返回进度
    for (int i = 0; i <= target; i++) {
        // 检查客户端是否取消请求
        if (as->isPreemptRequested() || !ros::ok()) {
            as->setPreempted();
            return;
        }

        // 设置进度反馈
        feedback.current = i;
        feedback.progress = (float)i / target;
        as->publishFeedback(feedback);
        ROS_INFO("Progress: current = %d, progress = %.2f", i, feedback.progress);

        rate.sleep();
    }

    // 执行完成，返回结果
    result.success = true;
    as->setSucceeded(result);
    ROS_INFO("Action finished successfully!");
}

int main(int argc, char **argv) {
    ros::init(argc, argv, "cpp_action_server");
    ros::NodeHandle nh;

    // 创建动作服务器（服务名：my_action）
    Server server(nh, "my_action", boost::bind(&execute, _1, &server), false);
    server.start();
    ROS_INFO("C++ Action Server started, waiting for goal...");

    ros::spin();
    return 0;
}
